### PR TITLE
fix(sequencer): validate dymint pubkey when creating sequencer

### LINKED ISSUE #708: Lack of public key validation when registering sequencer
The CreateSequencer function does not ensure the cryptographic validity of the provided public key. Specifically, there is no verification to confirm that the public key's point lies on the elliptic curve, which is a fundamental requirement for its validity in cryptographic operations.

### PR DESCRIPTION
**Pull request by srene**
## Description
According to the audit, the CreateSequencer function does not ensure the cryptographic validity of the provided public key. Specifically, there is no verification to confirm that the public key's point lies on the elliptic curve, which is a fundamental requirement for its validity in cryptographic operations.
This PR is aimed at validating that the public key is cryptographically correct using the IsOnCurve function.

----

Closes #708 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/sequencer/types/messages.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/decred/dcrd/dcrec/edwards"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
```
Comment (mtsitrin) [inline x/sequencer/types/messages.go:9]: what are these repos? why not using the tendermint ones?

File: x/sequencer/types/messages.go [hunk 2]
```diff
@@ -81,6 +84,22 @@ func (msg *MsgCreateSequencer) ValidateBasic() error {
...
 			return sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", pk)
 		}
+
+		var err error
+
+		// parse to secp256k1 or edc25519 pubkey and checks validity (the func includes cryptographic validation)
+		switch pk.Type() {
+		case "secp256k1":
+			_, err = secp256k1.ParsePubKey(pk.Bytes())
+		case "edc25519":
```
Comment (mtsitrin) [inline x/sequencer/types/messages.go:94]: only ed25519 expected for dymint key

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: x/sequencer/types/messages.go (hunk 1 above)
Comment (srene) [reply x/sequencer/types/messages.go:9]: afaik tendermint does not expose any validation function. what i did in this pr is to parse the keys to those libraries that will validate the public keys

File: x/sequencer/types/messages.go (hunk 2 above)
Comment (srene) [reply x/sequencer/types/messages.go:94]: i added both in case this changes in the future, but can be removed

**Review by omritoptix — APPROVED** [review]

**Review by zale144 — DISMISSED** [review]
lgtm

File: x/sequencer/types/messages.go (hunk 2 above)
Comment (mtsitrin) [reply x/sequencer/types/messages.go:94]: I think it should be removed

Comment (srene) [reply x/sequencer/types/messages.go]: i removed it, but then i think we should prevent creating sequencers with non ed25519 keys, which i think now is possible (but we should open a new issue for that)

**Review by mtsitrin — DISMISSED** [review]

File: x/sequencer/types/messages.go [hunk 3]
```diff
@@ -82,6 +83,13 @@ func (msg *MsgCreateSequencer) ValidateBasic() error {
 		if !ok {
 			return sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", pk)
 		}
+
+		_, err = edwards.ParsePubKey(edwards.Edwards(), pk.Bytes())
+		// err means the pubkey validation failed
+		if err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidPubKey, "pubkey validation failed")
```
Comment (danwt) [inline x/sequencer/types/messages.go:90]:
```suggestion
			return sdkerrors.Wrapf(sdkerrors.ErrInvalidPubKey, "%s", err)
```

**Review by danwt — CHANGES_REQUESTED** [review]

**Review by mtsitrin — DISMISSED** [review]

**Review by danwt — COMMENTED** [review]
Sorry if this isn't ready for review, I'm not sure
(If it's not, convert to draft maybe)
Is there a test for the thing we're trying to catch? (Using an invalid key)

File: x/sequencer/types/messages.go [hunk 4]
```diff
@@ -82,6 +83,13 @@ func (msg *MsgCreateSequencer) ValidateBasic() error {
 		if !ok {
 			return sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", pk)
 		}
+
+		_, err = edwards.ParsePubKey(edwards.Edwards(), pk.Bytes())
+		// err means the pubkey validation failed
+		if err != nil {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidPubKey, "%s", err)
```
[the remark this answers is not in the document]
Comment (srene) [reply x/sequencer/types/messages.go]: done

**Comment by srene** [comment]
> Sorry if this isn't ready for review, I'm not sure (If it's not, convert to draft maybe) Is there a test for the thing we're trying to catch? (Using an invalid key)

I added a test with an invalid key in TestMsgCreateSequencer_ValidateBasic

File: x/sequencer/types/messages.go [hunk 5]
```diff
@@ -67,29 +69,36 @@ func (msg *MsgCreateSequencer) GetSignBytes() []byte {
 func (msg *MsgCreateSequencer) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.Creator)
 	if err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+		return errorsmod.Wrapf(errortypes.ErrInvalidAddress, "invalid creator address (%s)", err)
 	}
 
 	// public key also checked by the application logic
 	if msg.DymintPubKey != nil {
```
Comment (mtsitrin) [inline x/sequencer/types/messages.go:76]: why we allow `DymintPubKey` to be nil in the msg?
we can set it to `nullable = false` in the proto file, to require it on the `MsgCreateSequencer`
I'll open separate issue for it

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**